### DISCUSSION TITLE
Ability to store additional custom fields on a rule?

### DISCUSSION THREAD

**Discussion by crhistianramirez** [react-querybuilder/react-querybuilder#611] [Q&A] [answered]
The data for one of my rules has a parent/child relationship. I'm using a custom value selector that allows me to select the parent resource, so that I can then select the child resource. The child ID is the value that needs to be saved to the rule for the expression but for display purposes it would be ideal to also be able to save the parentID (so that I can retrieve the user friendly names once the expression is saved). 

Is it possible to do this with the library? I tried using some other fields like title or testID but got an error that the object was not extensible.

**Comment by jakeboone02** [comment]
Short answer: yes. Longer answer: In version 7 this will be _much_ easier than in version 6 (the current `latest` version on npm). We have an alpha of v7 published ([`7.0.0-alpha.2`](https://www.npmjs.com/package/react-querybuilder/v/7.0.0-alpha.2), [migration instructions here](https://react-querybuilder.js.org/docs/next/migrate)), and it's probably safe for production to be honest. Would you like the longer/complicated answer for v6 or the shorter/easier answer for v7?

**Reply by crhistianramirez** [reply]
Thanks for the quick reply! I will take the shorter easier answer for v7

**Reply by jakeboone02** [reply, marked as answer]
[This sandbox](https://codesandbox.io/p/devbox/magical-cannon-dy7wmt?file=%2Fsrc%2FCustomValueEditor.tsx) should get you on the right track. [Relevant docs here](https://react-querybuilder.js.org/docs/next/migrate#query-selector-getter-and-dispatcher), relevant code below:

```tsx
import { ValueEditor, ValueEditorProps, update } from 'react-querybuilder';

export const CustomValueEditor = (props: ValueEditorProps) => {
  const handleOtherAttrChange = (val: string) => {
    const newQuery = update(
      props.schema.getQuery()!,
      'otherAttr' as any, // I need to fix this so "as any" isn't necessary
      val,
      props.path
    );
    props.schema.dispatchQuery(newQuery);
  };

  return (
    <div>
      <select
        value={(props.rule as any).otherAttr} // "as any" here because `RuleType` doesn't think `otherAttr` should be here
        onChange={e => handleOtherAttrChange(e.target.value)}>
        <option value="opt1">Option 1</option>
        <option value="opt2">Option 2</option>
        <option value="opt3">Option 3</option>
      </select>
      <ValueEditor {...props} />
    </div>
  );
};
```

**Reply by crhistianramirez** [reply]
Thanks! Those new query methods are going to simplify my code quite a bit

**Reply by crhistianramirez** [reply]
Btw, I am getting this error if I reload a page with react-query-builder on it:

> Error: request for 'react' is not in cache

The only dependency I updated is react-query-builder. Still looking into it, but I thought I'd let you know. If I can get to the bottom of it I will submit an issue

**Reply by jakeboone02** [reply]
Glad I could help!

> Btw, I am getting this error if I reload a page with react-query-builder on it:
> 
> > Error: request for 'react' is not in cache
> 
> The only dependency I updated is react-query-builder. Still looking into it, but I thought I'd let you know. If I can get to the bottom of it I will submit an issue

That's weird. What's your build system? If you find out it's an issue with RQB please file an issue.

**Reply by crhistianramirez** [reply]
@jakeboone02 its definitely something with my specific build/setup but I haven't narrowed down exactly what it is. I can't see it on the most recent version of nextjs so I think that will ultimately be the long-term fix but in the short term I think I will need to use the v6 version. To accomplish the same would it basically be passing down query as context, and then updating the query manually with the update utility function?

**Reply by jakeboone02** [reply]
> To accomplish the same would it basically be passing down query as context, and then updating the query manually with the update utility function?

Yeah, that's the gist of it. The part that can make it complicated is you have to manage state completely outside the component, including your update methods (don't use `handleOnChange`). I've got an example somewhere if you need me to search, but it sounds like you already have a pretty good idea of where to go.
